connectors/bigtable/flink-connector-gcp-bigtable/src/main/java/com/google/flink/connector/gcp/bigtable/table/BigtableDynamicTableSink.java

```diff
@@ -0,0 +1,144 @@
+
+import java.util.Objects;
+
+import static org.apache.flink.util.Preconditions.checkArgument;
+
+/**
+ * A {@link DynamicTableSink} for writing data into a Google Cloud Bigtable table. It generates the
+ * Sink for Table API.
+ */
+public class BigtableDynamicTableSink implements DynamicTableSink {
+    protected final Integer parallelism;
+    protected final ReadableConfig connectorOptions;
+    protected final ResolvedSchema resolvedSchema;
+    protected final String rowKeyField;
+
+    public BigtableDynamicTableSink(
+            ResolvedSchema resolvedSchema, ReadableConfig connectorOptions) {
+        checkArgument(
+                resolvedSchema.getPrimaryKeyIndexes().length == 1,
+                String.format(
+                        ErrorMessages.MULTIPLE_PRIMARY_KEYS,
+                        resolvedSchema.getPrimaryKeyIndexes().length));
+        int rowKeyIndex = resolvedSchema.getPrimaryKeyIndexes()[0];
+        checkArgument(
+                resolvedSchema
+                        .getColumn(rowKeyIndex)
+                        .get()
+                        .getDataType()
+                        .equals(DataTypes.STRING().notNull()),
+                String.format(
+                        ErrorMessages.ROW_KEY_STRING_TYPE_TEMPLATE,
+                        resolvedSchema.getColumn(rowKeyIndex).get().getDataType()));
+
+        this.connectorOptions = connectorOptions;
+        this.resolvedSchema = resolvedSchema;
+        this.parallelism = connectorOptions.get(BigtableConnectorOptions.SINK_PARALLELISM);
+        this.rowKeyField = resolvedSchema.getColumn(rowKeyIndex).get().getName();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(this.resolvedSchema, this.connectorOptions);
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) {
+            return true;
+        }
+        if (obj == null) {
+            return false;
+        }
+        if (getClass() != obj.getClass()) {
+            return false;
+        }
+        BigtableDynamicTableSink other = (BigtableDynamicTableSink) obj;
+        return Objects.equals(resolvedSchema, other.resolvedSchema)
+                && Objects.equals(connectorOptions, other.connectorOptions);
+    }
+
+    @Override
+    public ChangelogMode getChangelogMode(ChangelogMode requestedMode) {
+        return ChangelogMode.insertOnly();
+    }
+
+    @Override
+    public SinkRuntimeProvider getSinkRuntimeProvider(Context context) {
+        DataType physicalSchema = resolvedSchema.toPhysicalRowDataType();
+
+        RowDataToRowMutationSerializer.Builder serializerBuilder =
+                RowDataToRowMutationSerializer.builder()
+                        .withSchema(physicalSchema)
+                        .withRowKeyField(this.rowKeyField);
+
+        if (connectorOptions.get(BigtableConnectorOptions.USE_NESTED_ROWS_MODE)) {
+            serializerBuilder.withNestedRowsMode();
+        } else {
```

Comment:
should we log here if we're defaulting to column family mode?

Reply:
Users should know they are using column family as it needs to be set.